src/core/webview/webviewMessageHandler.ts

```diff
@@ -37,33 +37,38 @@ import { openMention } from "../mentions"
 import { telemetryService } from "../../services/telemetry/TelemetryService"
 import { TelemetrySetting } from "../../shared/TelemetrySetting"
 import { getWorkspacePath } from "../../utils/path"
-import { Mode, PromptComponent, defaultModeSlug, getModeBySlug, getGroupName } from "../../shared/modes"
+import { Mode, defaultModeSlug, getModeBySlug, getGroupName } from "../../shared/modes"
 import { getDiffStrategy } from "../diff/DiffStrategy"
 import { SYSTEM_PROMPT } from "../prompts/system"
 import { buildApiHandler } from "../../api"
+import { GlobalState } from "../../schemas"
 
 export const webviewMessageHandler = async (provider: ClineProvider, message: WebviewMessage) => {
+	// Utility functions provided for concise get/update of global state via contextProxy API.
+	const getGlobalState = <K extends keyof GlobalState>(key: K) => provider.contextProxy.getValue(key)
+	const updateGlobalState = async <K extends keyof GlobalState>(key: K, value: GlobalState[K]) =>
+		await provider.contextProxy.setValue(key, value)
+
 	switch (message.type) {
 		case "webviewDidLaunch":
 			// Load custom modes first
 			const customModes = await provider.customModesManager.getCustomModes()
-			await provider.updateGlobalState("customModes", customModes)
+			await updateGlobalState("customModes", customModes)
 
 			provider.postStateToWebview()
 			provider.workspaceTracker?.initializeFilePaths() // don't await
 
 			getTheme().then((theme) => provider.postMessageToWebview({ type: "theme", text: JSON.stringify(theme) }))
 
 			// If MCP Hub is already initialized, update the webview with current server list
-			if (provider.mcpHub) {
+			const mcpHub = provider.getMcpHub()
+			if (mcpHub) {
 				provider.postMessageToWebview({
 					type: "mcpServers",
-					mcpServers: provider.mcpHub.getAllServers(),
+					mcpServers: mcpHub.getAllServers(),
 				})
 			}
 
-			const cacheDir = await provider.ensureCacheDirectoryExists()
```

Comment:
This function is folded into the new higher-level `writeModelsToCache()` method.